jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/CoreSession.java

```diff
@@ -185,6 +196,47 @@ public interface CoreSession extends OutgoingFrames, Configuration
      */
     boolean isRsv3Used();
 
+    /**
+     * Used to notify the {@link CoreSession} that EOF has been read or the connection has been closed.
+     */
+    void onEof();
+
+    /**
+     * Used to notify the {@link CoreSession} that the connection has been opened.
+     */
+    void onOpen();
+
+    /**
+     * Process an Error that originated from the connection.
+     * For protocol causes, send and abnormal close frame
+     * otherwise just close the connection.
+     *
+     * @param cause the cause
+     * @param callback the callback on completion of error handling
+     */
+    void processConnectionError(Throwable cause, Callback callback);
+
+    /**
+     * Process an Error that originated from the handler.
+     * Send an abnormal close frame to ensure connection is closed.
+     *
+     * @param cause the cause
+     * @param callback the callback on completion of error handling
+     */
+    void processHandlerError(Throwable cause, Callback callback);
+
+    /**
+     * Set the class loader to be used by this {@link CoreSession}.
+     * @param classLoader the ClassLoader.
+     */
+    void setClassLoader(ClassLoader classLoader);
```

Comment:
In the server case we use context::run so it also calls the notifyEnterScope / notifyExitScope methods.

But I could do the client case with the `Consumer<Runnable>` constructor instead of setting the classloader.

Reply:
Not sure I follow.
I'm suggesting to change `from()` to take an additional `ClassLoader` as parameter, and if not necessary we pass `null`, saving a setter on an interface that is only ever used once.